File: GitToolBox/src/main/java/zielu/gittoolbox/status/GitAheadBehindCount.java
package zielu.gittoolbox.status;

import com.intellij.vcs.log.Hash;
import org.jetbrains.annotations.Nullable;

public class GitAheadBehindCount {
    public final RevListCount ahead;
    public final RevListCount behind;

    private GitAheadBehindCount(RevListCount _ahead, RevListCount _behind) {
        ahead = _ahead;
        behind = _behind;
    }

    private static GitAheadBehindCount create(RevListCount ahead, RevListCount behind) {
        return new GitAheadBehindCount(ahead, behind);
    }

    public static GitAheadBehindCount success(int ahead, @Nullable Hash aheadHash, int behind, @Nullable Hash behindHash) {
        return create(RevListCount.success(ahead, aheadHash), RevListCount.success(behind, behindHash));
    }

    public static GitAheadBehindCount cancel() {
        return create(RevListCount.cancel(), RevListCount.cancel());
    }

    public static GitAheadBehindCount failure() {
        return create(RevListCount.failure(), RevListCount.failure());
    }

    public boolean isNotZero() {
        if (status() == Status.Success) {
            return ahead.value() != 0 || behind.value() != 0;
        } else {
            return false;
        }
    }

    public boolean isNotZeroBehind() {
        if (status() == Status.Success) {
            return behind.value() != 0;
        } else {
            return false;
        }
    }

    public Status status() {
        return ahead.status();
    }

    @Override
    public String toString() {
        return ahead + " ahead, " + behind + " behind, status=" + status();
    }

    public static GitAheadBehindCount noRemote() {
        return create(RevListCount.noRemote(), RevListCount.noRemote());
    }
}

File: src/main/kotlin/zielu/gittoolbox/status/GitAheadBehindCount.kt
package zielu.gittoolbox.status

import com.intellij.vcs.log.Hash

internal data class GitAheadBehindCount constructor(
  val ahead: RevListCount,
  val behind: RevListCount
) {

  fun status(): Status = ahead.status()

  fun isNotZero(): Boolean {
    return if (status() == Status.SUCCESS) {
      ahead.value() != 0 || behind.value() != 0
    } else {
      false
    }
  }

  fun isNotZeroBehind(): Boolean {
    return if (status() == Status.SUCCESS) {
      behind.value() != 0
    } else {
      false
    }
  }

  companion object {
    private val cancel = GitAheadBehindCount(RevListCount.cancel(), RevListCount.cancel())
    private val failure = GitAheadBehindCount(RevListCount.failure(), RevListCount.failure())
    private val noRemote = GitAheadBehindCount(RevListCount.noRemote(), RevListCount.noRemote())

    @JvmStatic
    fun success(
      ahead: Int,
      aheadHash: Hash?,
      behind: Int,
      behindHash: Hash?
    ): GitAheadBehindCount {
      return GitAheadBehindCount(RevListCount.success(ahead, aheadHash), RevListCount.success(behind, behindHash))
    }

    @JvmStatic
    fun cancel(): GitAheadBehindCount = cancel

    @JvmStatic
    fun failure(): GitAheadBehindCount = failure

    @JvmStatic
    fun noRemote(): GitAheadBehindCount = noRemote
  }
}
